### PR TITLE
new PDF Importer for J&T Direktbank

### PR DESCRIPTION
**Pull request by ZfT2**
support  J&T Direktbank Tagesgeldkonto-Auszüge

### CODE REVIEW

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+{
+
+    private static final String DEPOSIT_2023 = "^(?<date>\\d+.\\d+.) (\\d+.\\d+.) (.*gutschr\\.?)(\\s*)(?<amount>[\\d\\s,.]*) [H]";
+    private static final String REMOVAL_2023 = "^(?<date>\\d+.\\d+.) (\\d+.\\d+.) (Überweisungsauftrag)(\\s*)(?<amount>[\\d\\s,.]*) [S]";
+    private static final String INTEREST_2023 = "^(?<date>\\d+.\\d+.) (\\d+.\\d+.) (Abschluss lt\\. Anlage \\d) (?<amount>[\\d\\s,.]*?) (.*)";
+
+    private static final String CONTEXT_KEY_YEAR = "year";
+    private static final String CONTEXT_KEY_CURRENCY = "currency";
+    private static final String CONTEXT_KEY_TRANSACTIONS_HAVE_FULL_DATE = "transactions_full_date";
+    private static final String CONTEXT_VALUE_TRUE = "true";
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java:27]: Hello @ZfT2 
thx... is it possible for you to adapt the [regular expression to our standard](https://github.com/portfolio-performance/portfolio/blob/master/CONTRIBUTING.md#regular-expressions)?

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    public JTDirektbankPDFExtractor(Client client)
+    {
+        super(client);
+
+        addBankIdentifier("500 150 01");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java:33]: Hello @ZfT2 
We cannot use a bank code for addBankIdentifier(). As soon as a user has this bank code, errors are generated.
Please use another one.
Also I don't see any test case which uses this addBankIdentifier().

Please remove it.

**Review by Nirus2000 — CHANGES_REQUESTED** [review]
Hey @ZfT2 , 
thanks for the contributions. 
It generally looks good to me. 
Have a look at the review comments I added.

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java (hunk 1 above)
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java:27]: Hello @ZfT2
thx... is it possible for you to adapt the [regular expression to our standard](https://github.com/portfolio-performance/portfolio/blob/master/CONTRIBUTING.md#regular-expressions)?

```suggestion
    private static final String DEPOSIT_2023 = "^(?<date>[\\d]{2}\\.[\\d]{2}\\.) [\\d]{2}\\.[\\d]{2}\\. (.*gutschr\\.?)[\\s]{1,}(?<amount>[\\.,\\d]+) [H]";
    private static final String REMOVAL_2023 = "^(?<date>[\\d]{2}\\.[\\d]{2}\\.) [\\d]{2}\\.[\\d]{2}\\. (.berweisungsauftrag)[\\s]{1,}(?<amount>[\\.,\\d]+) [S]";
    private static final String INTEREST_2023 = "^(?<date>[\\d]{2}\\.[\\d]{2}\\.) [\\d]{2}\\.[\\d]{2}\\. (Abschluss lt\\. Anlage \\d)[\\s]{1,}(?<amount>[\\.,\\d]+).*";

    private static final String CONTEXT_KEY_YEAR = "year";
    private static final String CONTEXT_KEY_CURRENCY = "currency";
    private static final String CONTEXT_KEY_TRANSACTIONS_HAVE_FULL_DATE = "transactions_full_date";
    private static final String CONTEXT_VALUE_TRUE = "true";
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        };
+    }
+
+    private BiConsumer<DocumentContext, String[]> contextProvider2023()
+    {
+        return (context, lines) -> {
+            Pattern yearPattern = Pattern.compile("(.*)(Kontoauszug Nr\\.)(\\s*)(\\d+)\\/(?<year>\\d{4})");
+            Pattern currencyPattern = Pattern
+                            .compile("(?<currency>[\\w]{3})(-Konto Kontonummer)(.*)");
+            contextProviderCommon(context, lines, yearPattern, currencyPattern);
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java:126]:
```suggestion
            Pattern yearPattern = Pattern.compile(".*Kontoauszug Nr\\.)[\\s]{1,}[\\d+]\\/(?<year>[\\d]{4})");
            Pattern currencyPattern = Pattern
                            .compile("(?<currency>[\\w]{3})\\-Konto Kontonummer.*");
            contextProviderCommon(context, lines, yearPattern, currencyPattern);
```

**Review by Nirus2000 — CHANGES_REQUESTED** [review]

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            transaction.setCurrencyCode(context.get(JTDirektbankPDFExtractor.CONTEXT_KEY_CURRENCY));
+        };
+    }
+
+    private BiConsumer<DocumentContext, String[]> contextProvider2023()
+    {
+        return (context, lines) -> {
+            Pattern yearPattern = Pattern.compile(".*Kontoauszug Nr\\.[\\s]{1,}[\\d+]\\/(?<year>[\\d]{4})");
+            Pattern currencyPattern = Pattern
+                            .compile("(?<currency>[\\w]{3})\\-Konto Kontonummer.*");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/JTDirektbankPDFExtractor.java:124]: Sorry, my fault. Here is the correct regex. 👍🏻 
```suggestion
            Pattern yearPattern = Pattern.compile("^.*Kontoauszug Nr\\.[\\s]{1,}[\\d]{1,2}\/(?<year>[\\d]{4})$");
            Pattern currencyPattern = Pattern.compile("^(?<currency>[\w]{3})\-Konto Kontonummer.*$");
```

**Review by Nirus2000 — CHANGES_REQUESTED** [review]

**Comment by ZfT2** [comment]
no problem, thanks for reviewing it :)

**Pull request merged**